hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py

```diff
@@ -0,0 +1,241 @@
+                if highest_level not in target_levels:
+                    target_levels.append(highest_level)
+            return target_levels
+        return [level for level in range(len(wsi_obj.level_downsamples))]
+
+    def get_level_data(self, wsi_obj: OpenSlide, level: int) -> np.ndarray:
+        """Returns data from a given level of the wsi pyramid. The data is returned in HWC format as expected by the
+        tiffwriter. The data must be in HWC format otherwise the compression will not work properly.
+
+        :param wsi_obj: The wsi object in openslide format
+        :param level: The level at which the data is extracted from the wsi pyramid
+        :return: A numpy array of shape (H, W, C) where H is the height, W is the width and C is the number of channels
+        """
+        level_data, _ = self.wsi_reader.get_data(wsi_obj, level=level)
+        level_data = level_data.transpose(1, 2, 0)
+        self.validate_level_data(level_data)
+        return level_data
+
+    def validate_level_data(self, level_data: np.ndarray) -> None:
+        """Makes sure that the level data is in HWC format. If the level data is not in HWC format, an error is raised.
+        This sanity check is necessary to avoid compression issues.
+        """
+        if level_data.shape[2] != 3:
+            raise ValueError(
+                f"Expected 3 channels but got {level_data.shape[2]}. Maybe the image is in channel first format? Try "
+                "transposing the image."
+            )
+
+    def get_options(self, wsi_obj: OpenSlide) -> Dict[str, Any]:
+        """Returns the options that will be passed to the tiffwriter. The options are extracted from the wsi properties
+        and will be written as tags in the tiff file.
+
+        :param wsi_obj: The wsi object in openslide format
+        :raises ValueError: Raises an error if the resolution unit is not in centimeters
+        :return: A dictionary of options that will be passed to the tiffwriter
+        """
+        resolution_unit = wsi_obj.properties[self.RESOLUTION_UNIT_KEY]
+
+        if resolution_unit != self.RESOLUTION_UNIT:
+            raise ValueError(f"Resolution unit is not in {self.RESOLUTION_UNIT}: {resolution_unit}")
+
+        options = dict(
+            software=self.SOFTWARE,
+            metadata={'axes': 'YXC'},
+            photometric=PHOTOMETRIC.RGB,
+            resolutionunit=resolution_unit,
+            compression=self.compression,
+            tile=(self.tile_size, self.tile_size),
+        )
+        return options
+
+    def get_px_per_cm_resolution_at_level(self, wsi_obj: OpenSlide, level: int) -> Tuple[float, float]:
+        """Returns the resolution of the wsi at a given level in pixels per centimeter.
+
+        :param wsi_obj: The wsi object in openslide format
+        :param level: The level at which the resolution is calculated
+        :return: A tuple of floats (x_resolution, y_resolution)
+        """
+        um_per_cm = 10000
```

Comment:
maybe add a check here the resolution unit is really centimetres?

Reply:
We check that in get_options from the properties